File: bridgepattern/bridgepattern.py
from abc import ABC, abstractmethod

class LivingThings(ABC):
    def __init__(self, breathe_implementor):
        self.breathe_implementor = breathe_implementor
    
    @abstractmethod
    def breathe_process(self):
        pass

class Dog(LivingThings):    
    def breathe_process(self):
        self.breathe_implementor.breathe()

class Fish(LivingThings):    
    def breathe_process(self):
        self.breathe_implementor.breathe()

class Tree(LivingThings):    
    def breathe_process(self):
        self.breathe_implementor.breathe()

class BreatheImplementor(ABC):
    @abstractmethod
    def breathe(self):
        pass

class LandBreatheImplementation(BreatheImplementor):
    def breathe(self):
        print('breathes through nose')

class WaterBreatheImplementation(BreatheImplementor):
    def breathe(self):
        print('breathes through gills')

class TreeBreatheImplementation(BreatheImplementor):
    def breathe(self):
        print('breathes through leaves')

if __name__ == '__main__':
    fish_obj = Fish(WaterBreatheImplementation())
    fish_obj.breathe_process()

    tree_obj = Tree(TreeBreatheImplementation())
    tree_obj.breathe_process()

File: bridgepattern/bridgepattern.js
// Abstract Implementor
class BreatheImplementor {
  breathe() {
    throw new Error('Method "breathe" must be implemented');
  }
}

// Concrete Implementors
class LandBreatheImplementation extends BreatheImplementor {
  breathe() {
    console.log('breathes through nose');
    console.log('inhale oxygen from air');
    console.log('exhale carbon dioxide');
  }
}

class WaterBreatheImplementation extends BreatheImplementor {
  breathe() {
    console.log('breathes through gills');
    console.log('inhale oxygen from water');
    console.log('exhale carbon dioxide');
  }
}

class TreeBreatheImplementation extends BreatheImplementor {
  breathe() {
    console.log('breathes through leaves');
    console.log('inhale carbon dioxide');
    console.log('exhale oxygen through photosynthesis');
  }
}

// Abstraction
class LivingThings {
  constructor(breatheImplementor) {
    if (new.target === LivingThings) {
      throw new Error("Cannot instantiate abstract class LivingThings directly");
    }
    this.breatheImplementor = breatheImplementor;
  }

  breatheProcess() {
    throw new Error('Method "breatheProcess" must be implemented');
  }
}

// Refined Abstractions
class Dog extends LivingThings {
  breatheProcess() {
    this.breatheImplementor.breathe();
  }
}

class Fish extends LivingThings {
  breatheProcess() {
    this.breatheImplementor.breathe();
  }
}

class Tree extends LivingThings {
  breatheProcess() {
    this.breatheImplementor.breathe();
  }
}

// Usage
const fish = new Fish(new WaterBreatheImplementation());
fish.breatheProcess();

console.log('---');

const tree = new Tree(new TreeBreatheImplementation());
tree.breatheProcess();
